### ISSUE TITLE
Web upload of large files fails with "too many open files"

### ISSUE THREAD

**Issue by srikumarks** [camlistore/camlistore#812]
The error I get on the `camlistored` side console is -

    2016/06/14 03:32:14 Sending error 500 to client for: writing to blobserver: open /Users/sriku/Library/Camlistore/blobs/sha1/af/6d/sha1-af6de74489f41a57e3d07d9b38cb30d2f4d4bd50.dat.tmp424234141: too many open files

This is for a multi-GB video file. The same file succeeds when using `camput`. This problem occurs on both MacOSX and Ubuntu.

**Comment by mpl** [comment]
@srikumarks Can you please give me some more info so I can try and reproduce ? such as:
- Camlistore revision
- approx size of the file (so I don't underestimate it)
- which browser
- devcam or production camlistored
- running camlistored remote or locally

**Comment by srikumarks** [comment]
@mpl - here you go -

* camlistored version: 2016-06-09-46cf811 (should I stick to releases/0.8?)
* I tried two files - 1.4GB and 3GB.
* Safari and Chrome
* Running remotely .. but on lan. Tried on localhost as well and faced the same issue.

**Comment by mpl** [comment]
@srikumarks thanks, that will help.

As to your question, it's entirely possible you're experiencing a bug we introduced since the last release, so feel free to retry with 0.9 if you want. However, we try to keep HEAD stable and bug free so there's usually not much advantage in sticking fervently with point releases imho.

**Comment by mpl** [comment]
@srikumarks oh, forgot to ask, your server-config.json too please (omit credentials if needed).

**Comment by mpl** [comment]
@srikumarks nevermind, I can reproduce. I'm on it.

**Comment by mpl** [comment]
@srikumarks I bet it's a ulimit problem. I don't get the errors anymore with https://camlistore-review.googlesource.com/6786 but it's a naive fix. You can either apply that fix, or set your ulimit for number of open files to a higher value while we think about a better fix.

@bradfitz I'm thinking we could either
- die with a useful error message when we get this kind of error (if we can detect it properly)
- or run ulimit -Sn during genconfig, or during the localdisk blobserver creation, and set a gate relevant with the value ulimit gives us.
WDYT?

**Comment by srikumarks** [comment]
@mpl Made a pull request 813 putting the ulimit into the code. sync guard pretty much does what ulimit does anyway. So this is ok. Besides the temp file opening code waits in a loop 1 second at a time till the file can be opened. This prevents failures in these kinds of cases near the ulimit instead of erroring out.

Thoughts?

**Comment by mpl** [comment]
@srikumarks 

- we don't do github PRs, see https://camlistore.org/code#contributing . In this case, we'll probably just iterate on https://camlistore-review.googlesource.com/6786 since that was the first CL proposed.
- I'm not sure we should set the ulimit ourselves, as it might go against a limit voluntarily set in place by the user/admin. Waiting on Brad's opinion on that.
- the 1s idle waiting seems a bit gross imho, especially since we have other options for this problem.

**Comment by srikumarks** [comment]
Sorry about the PR. Should've read the guidelines first.

Agree on the grossness of the wait, but it happens only on the extreme end of the limit and using the wait alone will cause the system to be robust at any ulimit, so the ulimit setting in the code can actually be discounted. It certainly doesn't have the failure conditions of the sync solution where the channel buffer length is determined once for the process lifetime. That works only as long as temp files are the only possible file handle hogging things in the server. Failing for such a simple use case is grosser :)

**Comment by mpl** [comment]
I agree about the locality limitation of the sync Gate (i.e. that it does not solve the problem if other processes are trying to get file handles as well at the same time), but the idle waiting has a similar shortcoming hasn't it? Sure, the tmp file itself will eventually succeed, but other processes trying the same thing at the same time elsewhere are going to fail.

Here's what I'm thinking: maybe we could have a more generic util func somewhere that takes a file opening operation/func as argument and repeats that operation until it succeeds. Except instead of doing it every second, maybe it should use an exponential back-off? Also, the first time it fails doing the operation, it should probably print a warning about it, mentioning that we end up retrying most likely because of a too low ulimit.
Then we could use that util not only in blobserver/localdisk , but in other places facing such risk.

WDYT? If you agree with the idea, do you want to open a new CL and implement it?

I'm not dismissing yet the idea of setting the ulimit ourselves, but I think the above proposal is kindof orthogonal and can be implemented regardless.

**Comment by srikumarks** [comment]
> Sure, the tmp file itself will eventually succeed, but other processes trying the same thing at the same time elsewhere are going to fail.

The problem with the sync gate is another one. Say we have the gate buffer at 1000, matching the current ulimit. If we have 100 non-tmp file handles already open and 900 tmp file handles already open, then the gate won't stop the 901's tmp file handle from being opened. Since that crosses the ulimit, the tmp open will fail, causing the non-extraordinary file upload operation to fail, which is undesirable.

With the 1 second loop scheme, at the ulimit, every tmp file open will just keep waiting for something else to close the handle. Given no handle leaks in goroutines, this has a fair chance of eventually succeeding.

> maybe we could have a more generic util func somewhere that takes a file opening operation/func as argument and repeats that operation until it succeeds.

Good idea.

> Except instead of doing it every second, maybe it should use an exponential back-off?

In this case, we *want* these operations to succeed and need to approach the handle-open optimistically. Ideally, I'd neither want to wait in a 1-second loop nor use exponential back-off. What is needed is a mechanism to be notified whenever opening a file handle will succeed next. Polling using exponential back-off could cause long gaps during which a file open that's started more recently suddenly gets to jump the queue over another that started earlier .. which is non-ideal. To an extent, this "queue jumping" problem also exists with the 1 second loop, but at least all tmp opens have roughly equal chance. We could put tmp file open requests into a queue if we *really* want to be fair, but the degree of fairness with the 1second loop is good enough I think.

The bottom line I think is this - if we know (or assume) that no goroutine leaks open handles, can we ensure that file uploads will eventually always succeed and not error out due to hitting the handle limit .. at least for ordinary cases?

> Also, the first time it fails doing the operation, it should probably print a warning about it, 
> mentioning that we end up retrying most likely because of a too low ulimit.
> Then we could use that util not only in blobserver/localdisk , but in other places facing such risk.

Good idea.

> WDYT? If you agree with the idea, do you want to open a new CL and implement it?

Happy to do that. Going through guidelines ..

**Comment by mpl** [comment]
@srikumarks ping. any progress? Should I take care of it instead?
